File: Data Structures & Algorithms/three-integer-sum/submission-2.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        for i in range(len(nums) - 2):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            s, e = i + 1, len(nums)-1
            while s < e:
                summ = nums[i] + nums[s] + nums[e]
                if summ == 0:
                    res.append([nums[i], nums[s], nums[e]])
                    while s < e and nums[s] == nums[s+1]:
                        s += 1
                    while s < e and nums[e] == nums[e-1]:
                        e -= 1
                    s += 1
                    e -= 1
                elif summ < 0:
                    s += 1
                else:
                    e -= 1
        return res

File: Data Structures & Algorithms/combination-target-sum/submission-2.rs
impl Solution {
    pub fn combination_sum(nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        fn dfs(
            nums: &[i32],
            idx: usize,
            rem: i32,
            cur: &mut Vec<i32>,
            res: &mut Vec<Vec<i32>>
        ) {
            if rem == 0 {
                res.push(cur.clone());
                return;
            }

            if rem < 0 || nums.len() == idx {
                return;
            }

            cur.push(nums[idx]);
            dfs(nums, idx, rem - nums[idx], cur, res);
            cur.pop();

            dfs(nums, idx+1, rem, cur, res);

        }

        let mut res = Vec::new();
        let mut cur = Vec::new();

        dfs(
            &nums,
            0,
            target,
            &mut cur,
            &mut res
        );

        res
    }
}
